### PR TITLE
[SPARK-42690][CONNECT] Implement CSV/JSON parsing functions for Scala client

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr add a new proto message 

```
message Parse {
  // (Required) Input relation to Parse. The input is expected to have single text column.
  Relation input = 1;
  // (Required) The expected format of the text.
  ParseFormat format = 2;

  // (Optional) DataType representing the schema. If not set, Spark will infer the schema.
  optional DataType schema = 3;

  // Options for the csv/json parser. The map key is case insensitive.
  map<string, string> options = 4;
  enum ParseFormat {
    PARSE_FORMAT_UNSPECIFIED = 0;
    PARSE_FORMAT_CSV = 1;
    PARSE_FORMAT_JSON = 2;
  }
}
```

and implement CSV/JSON parsing functions for Scala client.

### Why are the changes needed?
Add Spark connect jvm client api coverage.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass Github Actions
- Manual checked Scala 2.13

### CODE REVIEW

**Comment by zhengruifeng** [comment]
why not using `from_json` and `from_csv` to do this?

**Comment by LuciferYang** [comment]
> why not using `from_json` and `from_csv` to do this?

How to get the schema?

**Comment by zhengruifeng** [comment]
not sure whether I’m missing something, but isn’t the schema already provided by users?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 1]
```diff
@@ -644,6 +645,34 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
...
+  test("json from Dataset[String]") {
+    val session = spark
+    import session.implicits._
+    val expected = Seq(
+      new GenericRowWithSchema(
+        Array(73, "Shandong", "Kong"),
+        new StructType().add("age", LongType).add("city", StringType).add("name", StringType)))
+    val ds = Seq("""{"name":"Kong","age":73,"city":'Shandong'}""").toDS()
+    val result = spark.read.option("allowSingleQuotes", "true").json(ds)
+    checkSameResult(expected, result)
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:658]: https://github.com/apache/spark/blob/69dd20b5e45c7e3533efbfdc1974f59931c1b781/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L404-L417

From the code of the server side, `userSpecifiedSchema` is an `Option[StructType]` and default is None, so we can use it without specifying the`userSpecifiedSchema`?  Or is my test case not the correct scenario?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:658]: Thanks ~

Comment (zhengruifeng) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:658]: Probably we should add the user provided schema in the message? Or always discard it?

Comment (zhengruifeng) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:658]: Will inferFromDataset trigger an job? If so, I think we’d better skip it if possible

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala [hunk 2]
```diff
@@ -257,6 +258,13 @@ class PlanGenerationTestSuite
     session.read.json(testDataPath.resolve("people.json").toString)
   }
 
+  test("json from dataset") {
+    session.read
+      .schema(new StructType().add("c1", StringType).add("c2", IntegerType))
+      .option("allowSingleQuotes", "true")
+      .json(session.createDataset(Seq.empty[String])(StringEncoder))
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:265]: `session.emptyDataset(StringEncoder)`?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala [hunk 3]
```diff
@@ -131,7 +131,6 @@ object CheckConnectJvmClientCompatibility {
 
       // DataFrame Reader & Writer
       ProblemFilters.exclude[Problem]("org.apache.spark.sql.DataFrameReader.json"),
```
[the remark this answers is not in the document]
Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CheckConnectJvmClientCompatibility.scala:133]: There is no way we can support those, so this is the best we can do.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 4]
```diff
@@ -793,3 +794,23 @@ message CollectMetrics {
...
+message Parse {
+  // (Required) Input relation to Parse. The input is expected to have single text column.
+  Relation input = 1;
+  // (Required) The expected format of the text.
+  ParseFormat format = 2;
+
+  // (Optional) If not set, Spark will infer the schema.
+  //
+  // This schema string should be either DDL-formatted or JSON-formatted.
+  optional string schema = 3;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:807]: We can't use the actual data type?

File: connector/connect/common/src/test/resources/query-tests/explain-results/csv_from_dataset.explain [hunk 5]
```diff
@@ -0,0 +1 @@
+LogicalRDD [c1#0, c2#0], false
```
Comment (hvanhovell) [inline connector/connect/common/src/test/resources/query-tests/explain-results/csv_from_dataset.explain:1]: Oh this makes me sad. We are we using RDDs here?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 6]
```diff
@@ -710,6 +712,33 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+      localMap.foreach { case (key, value) => reader.option(key, value) }
+      reader
+    }
+    def ds: Dataset[String] = Dataset(session, transformRelation(rel.getInput))(Encoders.STRING)
+
+    rel.getFormat match {
+      case ParseFormat.PARSE_FORMAT_CSV =>
+        dataFrameReader.csv(ds).logicalPlan
+      case ParseFormat.PARSE_FORMAT_JSON =>
+        dataFrameReader.json(ds).logicalPlan
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:737]: Am I wrong here?

File: connector/connect/common/src/test/resources/query-tests/explain-results/csv_from_dataset.explain (hunk 5 above)
Comment (LuciferYang) [reply connector/connect/common/src/test/resources/query-tests/explain-results/csv_from_dataset.explain:1]: https://github.com/apache/spark/blob/39a55121888d2543a6056be65e0c74126a9d3bdf/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L424-L433

https://github.com/apache/spark/blob/39a55121888d2543a6056be65e0c74126a9d3bdf/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L503-L521

Comment (LuciferYang) [reply connector/connect/common/src/test/resources/query-tests/explain-results/csv_from_dataset.explain:1]: On the server side, the input `csvDataset` and `jsonDataset` are still `LocalRelation`, and the above code path is converted them to `RDD`.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
probably not related to this PR: 

https://github.com/apache/spark/blob/39a55121888d2543a6056be65e0c74126a9d3bdf/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L63-L76

```
  def schema(schemaString: String): DataFrameReader = {
    schema(StructType.fromDDL(schemaString))
  }
```

when the user provide a DDL string, it invoke the parser. Here I think we should keep both StructType and DDL string, and pass them to the server side.

**Comment by LuciferYang** [comment]
> probably not related to this PR:
> 
> https://github.com/apache/spark/blob/39a55121888d2543a6056be65e0c74126a9d3bdf/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L63-L76
> 
> ```
>   def schema(schemaString: String): DataFrameReader = {
>     schema(StructType.fromDDL(schemaString))
>   }
> ```
> 
> when the user provide a DDL string, it invoke the parser. Here I think we should keep both StructType and DDL string, and pass them to the server side.

message `Read` seems also need to consider this？I think we can further discuss this problem in a separate jira?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 7]
```diff
@@ -504,6 +542,18 @@ class DataFrameReader private[sql] (sparkSession: SparkSession) extends Logging
     }
   }
 
+  private def parse(ds: Dataset[String], format: ParseFormat): DataFrame = {
+    sparkSession.newDataFrame { builder =>
+      val parseBuilder = builder.getParseBuilder
+        .setInput(ds.plan.getRoot)
+        .setFormat(format)
+      userSpecifiedSchema.foreach(schema => parseBuilder.setSchema(schema.toDDL))
```
Comment (zhengruifeng) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:550]: as to this PR itself, I think we should probably use `DataType schema` in the proto message, `schema.toDDL` always discards the metadata.

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:550]: What do you think about this? @hvanhovell

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: [20f1722](https://github.com/apache/spark/pull/40332/commits/20f1722c4cc3331e0c8247f4d3cae419a3e81870) change to pass a DataType @zhengruifeng

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: [ca6ec7b](https://github.com/apache/spark/pull/40332/commits/ca6ec7b4651b3e9c69bc96240687840179b49ff0) rename `data_type` to `schema` in proto message

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: In this scenario, I think `DataType schema` is used and `schema.toDDL` is no longer needed

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
@LuciferYang thank you for woking on this.

merged into master/branch-3.4

**Comment by LuciferYang** [comment]
Thanks @zhengruifeng @hvanhovell @HyukjinKwon  ~

**Pull request closed without merging**